### PR TITLE
list objects pipe enhancement

### PR DESCRIPTION
**Pull request by senojj**
## Description
This PR introduces a change in the implementation of the pipe that underpins the list objects pipeline algorithm. The change is intended to increase performance and overall efficiency of a pipe.
#### What problem is being solved?
Currently, the pipe keeps track of its size in a variable that is updated, behind a mutex, by each thread that is sending and receiving messages on the pipe. This is bad for cache locality.
#### How is it being solved?
The `count` field of a `Pipe` is removed, leaving only the `head` field to be updated by sending threads, and the `tail` field to be updated by receiving threads.

Additionally, the `head` and `tail` fields are now only incremented. Their values were previously altered via modular arithmetic, wrapping after reaching the size of the buffer. Instead, the values are now masked with a bitwise AND to perform the wrapping. When the field value becomes too large to fit into a `uint`, the value will wrap back to 0.

To support these optimizations, the buffer size must now be a valid power of two. This is necessary to support the integer overflow wrapping, and bitwise masking. This technique removes the need to track the current size of the buffer in another field.

There is now no conventional way to support a "grow" operation, so the `Pipe.Grow` method has been removed from the type.

The size must now be configured at time of instantiation, so a `New` function was introduced to instantiate a Pipe. Now that initialization happens at time of instantiation, the initialize helper method, and the `init` field have been removed from the Pipe.

All of the above changes have resulted in a smaller Pipe memory footprint, and a noticeable decrease in ns/op in benchmarks.

## Review Checklist
- [ ] I have clicked on ["allow edits by maintainers"](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [ ] I have added documentation for new/changed functionality in this PR or in a PR to [openfga.dev](https://github.com/openfga/openfga.dev) [Provide a link to any relevant PRs in the references section above]
- [ ] The correct base branch is being used, if not `main`
- [ ] I have added tests to validate that the change in functionality is working as expected

## Summary by CodeRabbit

* **Refactor**
  * Restructured internal messaging pipe implementation to improve performance and reliability.
  * Enhanced buffer management with stricter validation requirements and optimized state tracking.
  * Updated pipeline initialization to use consistent defaults and validation for configuration parameters.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: internal/pipe/messaging.go [hunk 1]
```diff
@@ -20,39 +20,45 @@ type TxCloser[T any] interface {
...
+		panic("value provided to Pipe.Grow must be a power of 2")
 	}
-	p.mu.Lock()
-	defer p.mu.Unlock()
+	var p Pipe[T]
+	p.data = make([]T, n)
+	p.condFull = sync.NewCond(&p.mu)
+	p.condEmpty = sync.NewCond(&p.mu)
+	return &p
+}
```
[the remark this answers is not in the document]
Comment (senojj) [reply internal/pipe/messaging.go:47]: This issue is address in commit b662fdc

**Comment by senojj** [comment]
@claude

File: internal/pipe/messaging.go [hunk 2]
```diff
@@ -73,78 +79,60 @@ func (p *Pipe[T]) Seq() iter.Seq[T] {
...
 
 	if p.done {
 		return false
 	}
 
-	p.data[p.head] = item
-	p.head = (p.head + 1) % len(p.data)
-	p.count++
+	p.head++
+	p.data[p.mask(p.head)] = item
```
[the remark this answers is not in the document]
Comment (senojj) [reply internal/pipe/messaging.go:96]: This is addressed in commit 4e9d4f4

File: pkg/server/commands/reverseexpand/pipeline/pipeline.go [hunk 3]
```diff
@@ -54,16 +54,26 @@ var (
...
 
+const (
+	defaultBufferSize int = 128
+	defaultChunkSize  int = 100
+	defaultNumProcs   int = 3
+)
+
+func powerOfTwo(n int) bool {
+	return n > 0 && (n&(n-1)) == 0
+}
```
[the remark this answers is not in the document]
Comment (senojj) [reply pkg/server/commands/reverseexpand/pipeline/pipeline.go]: Addressed in commit afb9984

**Comment by senojj** [comment]
@claude

File: internal/pipe/messaging.go [hunk 4]
```diff
@@ -73,78 +77,60 @@ func (p *Pipe[T]) Seq() iter.Seq[T] {
...
 
 	if p.done {
 		return false
 	}
 
-	p.data[p.head] = item
-	p.head = (p.head + 1) % len(p.data)
-	p.count++
+	p.data[p.mask(p.head)] = item
+	p.head++
```
[the remark this answers is not in the document]
Comment (senojj) [reply internal/pipe/messaging.go:97]: This was addressed

File: internal/pipe/messaging.go [hunk 5]
```diff
@@ -20,39 +22,41 @@ type TxCloser[T any] interface {
...
-	defer p.mu.Unlock()
+	var p Pipe[T]
+	p.data = make([]T, n)
+	p.condFull = sync.NewCond(&p.mu)
+	p.condEmpty = sync.NewCond(&p.mu)
+	return &p
+}
 
-	replacement := make([]T, len(p.data)+n)
+func (p *Pipe[T]) size() uint {
```
[the remark this answers is not in the document]
Comment (senojj) [reply internal/pipe/messaging.go]: addressed

File: pkg/server/commands/reverseexpand/pipeline/pipeline.go [hunk 6]
```diff
@@ -74,9 +81,13 @@ func New(backend *Backend, options ...Option) *Pipeline {
...
 }
 
+// WithBufferSize is a function that sets the value for a pipeline's workers' internal
+// pipe buffer size. The value must be a power of two. (e.g. 1, 2, 4, 8, 16, 32...)
+// The default value of the buffer size is 128. If an invalid value is provided as size
+// then the default value will be applied.
 func WithBufferSize(size int) Option {
-	if size < 1 {
-		size = 1
+	if !bits.PowerOfTwo(size) {
```
[the remark this answers is not in the document]
Comment (senojj) [reply pkg/server/commands/reverseexpand/pipeline/pipeline.go]: this is known.

File: internal/pipe/messaging_test.go
Comment (adriantam) [inline internal/pipe/messaging_test.go:1]: Looking at the [internal/pipe/messaging.go](https://github.com/openfga/openfga/pull/2853/files#diff-80d17023a0b727023f408caedfa78f530438585f81a2d5fa2d9c2ad6ea1a33fb) shows the code should work.  However, for completeness sake, we should have test where the number of times Send and Recv is greater than the pipe size, this will test whether the circular logic works correctly.  

Also, do we have tests (other than benchmark) where the pipe is full / empty?

**Review by yissellokta — APPROVED** [review]

File: internal/pipe/messaging_test.go
Comment (senojj) [reply internal/pipe/messaging_test.go:1]: > However, for completeness sake, we should have test where the number of times Send and Recv is greater than the pipe size, this will test whether the circular logic works correctly.

This already exists in `internal/pipe/messaging_test.go`.

> Also, do we have tests (other than benchmark) where the pipe is full / empty?

Yes, in the same place as noted above.

Note that a Pipe is blocking when attempting to send to a Pipe that is at capacity, and when attempting to receive from a Pipe that is empty. That is why the tests produce to and consume from the Pipe in separate goroutines.

Additional note: Many more tests can be created to more comprehensively test the Pipe type; are they necessary to move this PR forward?
